src/Items/ItemLilypad.h

```diff
@@ -57,51 +51,146 @@ class cItemLilypadHandler :
+			0, 0, 0
+		);
+	}
+
+	virtual bool OnPlayerPlace(
+        cWorld & a_World, cPlayer & a_Player, const cItem & a_EquippedItem,
+        int a_BlockX, int a_BlockY, int a_BlockZ, eBlockFace a_BlockFace,
+        int a_CursorX, int a_CursorY, int a_CursorZ
+    ) override
+	{
+		Vector3i PlacementPos = GetBlockNextTo(Vector3i(a_BlockX, a_BlockY, a_BlockZ), a_BlockFace);
+		
+		// If the block where we are trying to place is a water block, try placing on top of the water block instead
+		BLOCKTYPE Placement = a_World.GetBlock(PlacementPos);
+		NIBBLETYPE PlacementMeta = a_World.GetBlockMeta(PlacementPos);
+		if(((Placement == E_BLOCK_STATIONARY_WATER) || (Placement == E_BLOCK_WATER)) && (PlacementMeta == 0))  // A water source is in the place of where we want to place the block
+		{
+			return false;
+		}	
+
+		BLOCKTYPE Surface = a_World.GetBlock(PlacementPos - Vector3i(0, 1, 0));
+		NIBBLETYPE SurfaceMeta = a_World.GetBlockMeta(PlacementPos - Vector3i(0, 1, 0));
+		if(!((((Surface == E_BLOCK_STATIONARY_WATER) || (Surface == E_BLOCK_WATER)) && (SurfaceMeta == 0)) ||  // A water source is below
+			(Surface == E_BLOCK_ICE) || (Surface == E_BLOCK_FROSTED_ICE)))                                     // Or (frosted) ice
 		{
-			a_World->SetBlock(Callbacks.m_Pos.x, Callbacks.m_Pos.y, Callbacks.m_Pos.z, E_BLOCK_LILY_PAD, 0);
-			if (!a_Player->IsGameModeCreative())
+			return false;
+		}
+
+		// Otherwise do the same exact process as cItemHandler::OnPlayerPlace, but ignore build collision
+
+		Vector3i absPos(a_BlockX, a_BlockY, a_BlockZ);
+
+		if (a_BlockFace < 0)
+		{
+			// Clicked in air
+			return false;
+		}
+	
+		if (!cChunkDef::IsValidHeight(absPos.y))
+		{
+			// The clicked block is outside the world, ignore this call altogether (#128)
+			return false;
+		}
+	
+		BLOCKTYPE ClickedBlock;
+		NIBBLETYPE ClickedBlockMeta;
+		a_World.GetBlockTypeMeta(absPos.x, absPos.y, absPos.z, ClickedBlock, ClickedBlockMeta);
+
+		cChunkInterface ChunkInterface(a_World.GetChunkMap());
+		
+		absPos = GetBlockNextTo(absPos, a_BlockFace);
+
+		if (!cChunkDef::IsValidHeight(absPos.y))
+		{
+			// The block is being placed outside the world, ignore this packet altogether (#128)
+			return false;
+		}
+
+		NIBBLETYPE PlaceMeta;
+		BLOCKTYPE PlaceBlock;
+		a_World.GetBlockTypeMeta(absPos.x, absPos.y, absPos.z, PlaceBlock, PlaceMeta);
+
+		// Check for another block here, remove it if appropriate.
+		if (!BlockHandler(PlaceBlock)->DoesIgnoreBuildCollision(ChunkInterface, absPos, a_Player, PlaceMeta))
+		{
+			// Tried to place a block into another?
+			// Happens when you place a block aiming at side of block with a torch on it or stem beside it
+			return false;
+		}
+
+		a_World.DropBlockAsPickups(absPos, &a_Player, nullptr);
+	
+		// Get all the blocks to place:
+		sSetBlockVector blocks;
+		if (!GetBlocksToPlace(a_World, a_Player, a_EquippedItem, absPos.x, absPos.y, absPos.z, a_BlockFace, a_CursorX, a_CursorY, a_CursorZ, blocks))
```

Comment:
No point in calling `GetBlocksToPlace`. We know it will be a lily pad.

Just call `a_Player.PlaceBlock(...)` directly.

Reply:
That seems too hard-coded?

Reply:
An item handler's entire purpose is to "hard-code" the behaviour of that one specific item. Making it more complicated than necessary just makes it harder to understand the code when coming back to it later.